### ISSUE TITLE
proposal - c7n-test - policy testing

### ISSUE THREAD

**Issue by marcoceppi** [cloud-custodian/cloud-custodian#6407]
**Is your feature request related to a problem? Please describe.**
The proliferation of policies created in organizations can quickly be overwhelming to maintain and continue to produce. Having a high degree of confidence that modifications made to policy filters and actions produce desired results feels imperative to custodian's continued adoption.

**Describe the solution you'd like**
I'd like to introduce a new c7n tool, `c7n-test`, and framework which policy authors can validate that policies perform against the right set of resources and execute the correct actions. The full proposal, which is a YAML DSL for describing test parameters and a CLI tool for executing them is documented in full on this [Google Doc: Policy Testing Proposal](https://docs.google.com/document/d/1-MKGbzXi-vxKbYLIUQ9kvOa_7KYcHEnWF8FhZWhqFjc/edit#).

In summary, a tool that can execute infrastructure as code (terraform, cloud formations, kubernetes manifests, etc) to stand up both compliant and non-compliant resources that policies are then executed against and the results evaluated. The CLI tool is designed to be extensible and pluggable to accept a large range of IaC tools to support the large range of what an organization or group may prefer. To start, I suggest focusing on terraform but the solution will not be limited to any one.

**Additional context**
- Inspired by and related to #6023
- Related to, but not the same as, [functional testing for c7n](https://cloudcustodian.io/docs/developer/tests.html#writing-tests-for-cloud-controlled-resources)

**Comment by JohnHillegass** [comment]
Awesome take in this idea @marcoceppi 👍  If there is any way that I can, I'd love to help with this!

**Comment by matticulous** [comment]
I like a good portion of the proposal -- particularly the pluggable backends for infrastructure setup. I'm mixed on the removal of replay. I very much appreciate focusing on an MVP, so completely understand. We are trying to do functional testing of policies following the methods described [here](https://cloudcustodian.io/docs/developer/tests.html), albeit in our own policy repo. This has been helpful in trying to ensure testing is reasonably consistent (e.g. context setup, coverage) b/t policy authors. We use replay for two main purposes: 1) refine fixtured data to resources under test (i.e. operating in an environment/account with varying amount of unrelated resources), 2) executing tests in a CI pipeline. No matter the testing mechanic, I think it is good to identify its intent and/or scope.

**Comment by marcoceppi** [comment]
Thanks @matticulous for taking time to read over and provide feedback. I've been on the fence about flight recordings. I'm a big fan of them in this repo because they speed up an otherwise lengthy task of phoning to a real cloud. They work well here because the code under test shouldn't change drastically enough that a fresh cloud run is required except for special circumstances.

Where I'm stuck on flight recording for policy testing: any change, substantial or otherwise, to a c7n policy will likely require a re-run against live test targets to validate. Flight recordings also only really work with pull mode policies. A framework for replaying cloud events for validation or handling those hasn't been solved yet for c7n testing and would be another limitation.

To address your two points:

> 1) refine fixtured data to resources under test (i.e. operating in an environment/account with varying amount of unrelated resources)

To combat this, and blast radius in general, `c7n-test` will instrument some strict checks against which resources it'll filter/match and ultimately remediate. Each engine will return the list of ARNs/identifiers for the resources created and those will be implicitly added as an AND filter on the policy under test. This way any filter rules will ONLY match those resources created specifically for the test case. This lowers blast radius of accidental matches if resources exist in the account under test and allows us to run `c7n-test` in parallel mode for faster executions in the same account.

> 2) executing tests in a CI pipeline. No matter the testing mechanic, I think it is good to identify its intent and/or scope.

So this I could see the value in this, especially if you're trying to test things outside of a policy file change like an upgrade to c7n itself: Given a set of cloud resources as inputs does c7n filter and action them as expected from a prior version. This use case still feels like there's _more_ value in just running those same tests live. The CLI itself will be purpose built with CI systems in mind making sure that output handling, inputs from flags/environment variables, and sniffing out common CI systems to enable enhanced interactions.

You're not the first person to mention flight recordings so there seems to be an appetite for it. As such, having a feature like flight recordings in `c7n-test` might be a good idea. I'm not sure what shape it would take but I'll update the proposal so as c7n-test is  designed more, and ultimately implemented, we will consider what hooks would be good to instrument some kind of record/playback mode across all engines.
